### PR TITLE
lastActive

### PR DESCRIPTION
**Pull request by ritishgumber**

### CODE REVIEW

File: api/project.js [hunk 1]
```diff
@@ -1,399 +1,489 @@
...
+            });
+
+        } else {
+            console.log("Missing appId in the parameter.");
+            return res.send(401);
+        }
+
+    });
+
+    app.delete('/app/inactive', function(req, res, next) {
```
Comment (nawazdhandala) [inline api/project.js]: change this api to apps/deleteInactive

File: api/project.js [hunk 2]
```diff
@@ -1,399 +1,489 @@
...
+
+        console.log("App Active API");
+
+        global.projectService.deleteInactiveApps().then(function(inactiveApps) {
+            if (!inactiveApps) {
```
Comment (nawazdhandala) [inline api/project.js]: !inactiveApps -> Does this mean an error, is yes, Handle this in the error callback.

File: api/project.js [hunk 3]
```diff
@@ -1,399 +1,489 @@
...
+            return res.status(200).send(inactiveApps);
+
+        }, function(error) {
+            console.log("Error deleting inactiveApps.");
+            return res.status(500).send(error);
+        });
+
+    });
+
+    app.post('/app/notifyInactive', function(req, res, next) {
```
Comment (nawazdhandala) [inline api/project.js]: apps/notifyInactive

File: api/project.js [hunk 4]
```diff
@@ -1,399 +1,489 @@
...
+
+        console.log("Notify Inactive Apps");
+        global.projectService.notifyInactiveApps().then(function(inactiveApps) {
+            if (!inactiveApps) {
```
Comment (nawazdhandala) [inline api/project.js]: You cannot have an error in success callback.

File: model/project.js [hunk 5]
```diff
@@ -1,19 +1,20 @@
...
+        appId: String,
+        _userId: String,
+        keys: Object,
+        invited: Array,
+        developers: Array,
+        planId: Number,
+        provider: Object,
+        providerProperties: Object,
+        disabled: Boolean,
+        lastActive: Number
```
Comment (nawazdhandala) [inline model/project.js]: Should be date time.

File: helpers/utils.js [hunk 6]
```diff
@@ -1,27 +1,36 @@
...
+};
+
+function loadTemplate(file, data) {
+    return ejs.render(templates[file], data);
+}
+
+function slugify(text) {
+    return text.replace(/[^\w]/g, '-').toLowerCase();
+}
+function _request(method, url, data) {
```
Comment (nawazdhandala) [inline helpers/utils.js]: What does this do?

File: api/project.js (hunk 1 above)
Comment (nawazdhandala) [inline api/project.js]: Secure this API with secureKey

File: api/project.js (hunk 3 above)
Comment (nawazdhandala) [inline api/project.js]: Secure this APi with secureKey

File: services/projectService.js [hunk 7]
```diff
@@ -1,1540 +1,1818 @@
...
+                var inactiveApps = [];
+                Project.find({}, function(err, projects) {
+                    if (err) {
+                        console.log("Error in Getting projects...");
+                        deferred.reject(err);
+                    } else {
+                        var length = projects.length;
+                        projects.forEach(function(project, index) {
+                            length--;
+                            if (Date.now() - project._doc.lastActive > 5184000000) {
```
Comment (nawazdhandala) [inline services/projectService.js]: Comment on how much time is this. :)

File: services/projectService.js [hunk 8]
```diff
@@ -1,1540 +1,1818 @@
...
+                                inactiveApps.push(project._doc.appId);
+                                User.findById(project._doc._userId, function(err, user) {
+                                    global.mailService.sendTextMail(keys.adminEmailAddress, user._doc.email, "Inactive App", "Its been more than 60 days .").then(function(info) {
```
Comment (nawazdhandala) [inline services/projectService.js]: Have a good email with a template, also have the link in the email to activate your app. Have a good text in the email. Check "Activate Account" email for the template.

File: services/projectService.js [hunk 9]
```diff
@@ -1,1540 +1,1818 @@
...
+                                            deferred.resolve(inactiveApps);
+                                        }
+                                    , function(err) {
+                                        deferred.reject(err);
+                                    });
+
+                                })
+                            } else {
+                                if (length == 0)
+                                    deferred.resolve(inactiveApps);
```
Comment (nawazdhandala) [inline services/projectService.js]: Return empty array in the promise []. Makes it easy to read.

File: services/projectService.js [hunk 10]
```diff
@@ -1,1540 +1,1818 @@
...
+                var inactiveApps = [];
+                Project.find({}, function(err, projects) {
+                    if (err) {
+                        console.log("Error in Getting projects...");
+                        deferred.reject(err);
+                    } else {
+                        var length = projects.length;
+                        projects.forEach(function(project) {
+                            length--;
+                            //7776000000
```
Comment (nawazdhandala) [inline services/projectService.js]: Comment on how much time is this.

File: services/projectService.js [hunk 11]
```diff
@@ -1,1540 +1,1818 @@
...
+                            if (Date.now() - project._doc.lastActive > 7776000000) {
+                                inactiveApps.push(project._doc.appId);
+                                User.findById(project._doc._userId, function(err, user) {
+                                    if (err)
+                                        deferred.reject(err);
+                                    else {
+                                        global.mailService.sendTextMail(keys.adminEmailAddress, user._doc.email, "Delete App", "Its been more than 90 days,thus deleting your app .").then(function(info) {
```
Comment (nawazdhandala) [inline services/projectService.js]: Have a good email template.

File: api/project.js (hunk 1 above)
Comment (ritishgumber) [inline api/project.js]: the request is already of type delete

File: helpers/utils.js (hunk 6 above)
Comment (ritishgumber) [inline helpers/utils.js]: It was already there , I only created a _request function

File: services/projectService.js (hunk 9 above)
Comment (ritishgumber) [inline services/projectService.js]: it returns the array of inactive appIds

File: api/project.js (hunk 1 above)
Comment (nawazdhandala) [inline api/project.js]: Oops, did not check that. Sorry.

File: services/projectService.js (hunk 9 above)
Comment (nawazdhandala) [inline services/projectService.js]: ok

**Pull request merged**